File: ghidra_scripts/function_export.py
# Export Ghidra functions to JSON for AI Ingestion
# @author re-Brain
# @category Analysis
# @keybinding 
# @menupath 
# @toolbar 

import json
import os
import json
import os
# from ghidra.app.util.headless import HeadlessAnalyzer

def run():
    program = getCurrentProgram()
    fm = program.getFunctionManager()
    functions = fm.getFunctions(True) # True for forward iteration
    
    export_data = {
        "project": program.getDomainFile().getProjectLocator().getName(),
        "binary": program.getName(),
        "functions": []
    }
    
    for f in functions:
        func_info = {
            "name": f.getName(),
            "address": f.getEntryPoint().toString(),
            "signature": f.getPrototypeString(True, True),
            "comment": f.getComment()
        }
        export_data["functions"].append(func_info)
        
    # Standard export path for the watcher to pick up
    export_path = "/ghidra/jobs/export.json"
    with open(export_path, "w") as f:
        json.dump(export_data, f, indent=2)
    
    print("Successfully exported {} functions to {}".format(len(export_data["functions"]), export_path))

if __name__ == "__main__":
    try:
        run()
    except Exception as e:
        import traceback
        traceback.print_exc()
        print(f"Export Script Error: {e}")

File: ghidra_scripts/FunctionExport.java
//Export functions to JSON
//@author re-Brain
//@category Analysis
//@keybinding
//@menupath
//@toolbar

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import java.io.FileWriter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

public class FunctionExport extends GhidraScript {
    @Override
    public void run() throws Exception {
        if (currentProgram == null) {
            println("No current program.");
            return;
        }

        FunctionManager fm = currentProgram.getFunctionManager();
        JsonObject exportData = new JsonObject();
        exportData.addProperty("project", currentProgram.getDomainFile().getProjectLocator().getName());
        exportData.addProperty("binary", currentProgram.getName());
        
        JsonArray functions = new JsonArray();
        for (Function f : fm.getFunctions(true)) {
            JsonObject funcInfo = new JsonObject();
            funcInfo.addProperty("name", f.getName());
            funcInfo.addProperty("address", f.getEntryPoint().toString());
            funcInfo.addProperty("signature", f.getPrototypeString(true, true));
            if (f.getComment() != null) {
                funcInfo.addProperty("comment", f.getComment());
            }
            functions.add(funcInfo);
        }
        exportData.add("functions", functions);
        
        String exportPath = "/ghidra/jobs/export.json";
        try (FileWriter writer = new FileWriter(exportPath)) {
            new GsonBuilder().setPrettyPrinting().create().toJson(exportData, writer);
        }
        println("Successfully exported " + functions.size() + " functions to " + exportPath);
    }
}
